File: core/src/no/elg/infiniteBootleg/world/box2d/ChunkBody.java
package no.elg.infiniteBootleg.world.box2d;

import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.physics.box2d.Body;
import com.badlogic.gdx.physics.box2d.BodyDef;
import com.badlogic.gdx.physics.box2d.EdgeShape;
import com.badlogic.gdx.physics.box2d.Fixture;
import com.badlogic.gdx.utils.Disposable;
import no.elg.infiniteBootleg.Main;
import no.elg.infiniteBootleg.util.CoordUtil;
import no.elg.infiniteBootleg.util.Tuple;
import no.elg.infiniteBootleg.world.*;
import no.elg.infiniteBootleg.world.render.WorldRender;
import org.jetbrains.annotations.NotNull;

import static no.elg.infiniteBootleg.world.Chunk.CHUNK_SIZE;

/**
 * @author Elg
 */
public class ChunkBody implements Disposable {

    private final Chunk chunk;
    private Body box2dBody;

    private final static Tuple<Direction, byte[]>[] EDGE_DEF;

    static {
        //represent the direction to look and if no solid block there how to create a fixture at that location (ie
        // two relative vectors)
        // the value of the tuple is as follows dxStart, dyStart, dxEnd, dyEnd
        // this can be visually represented with a cube:
        //
        // (0,1)---(1,1)
        //   |       |
        //   |       |
        //   |       |
        // (0,0)---(1,0)
        //
        // where 'd' stands for delta
        // x/y is if this is the x or component of the coordinate
        // end/start is if this is the start or end vector
        //noinspection unchecked
        EDGE_DEF = new Tuple[4];
        EDGE_DEF[0] = new Tuple<>(Direction.NORTH, new byte[] {0, 1, 1, 1});
        EDGE_DEF[1] = new Tuple<>(Direction.EAST, new byte[] {1, 0, 1, 1});
        EDGE_DEF[2] = new Tuple<>(Direction.SOUTH, new byte[] {0, 0, 1, 0});
        EDGE_DEF[3] = new Tuple<>(Direction.WEST, new byte[] {0, 0, 0, 1});
    }

    public ChunkBody(@NotNull Chunk chunk) {
        this.chunk = chunk;
    }

    /**
     * Update the box2d fixture of this block
     *
     * @param recalculateNeighbors
     *     If the neighbors also should be updated
     */
    public void updateFixture(boolean recalculateNeighbors) {
        if (chunk.isAllAir()) {
            if (box2dBody != null) {
                synchronized (WorldRender.BOX2D_LOCK) {
                    chunk.getWorld().getBox2dWorld().destroyBody(box2dBody);
                }
            }
            return;
        }

        //recalculate the shape of the chunk (box2d)

        BodyDef bodyDef = new BodyDef();
        bodyDef.position.set(chunk.getChunkX() * CHUNK_SIZE, chunk.getChunkY() * CHUNK_SIZE);
        bodyDef.fixedRotation = true;
        bodyDef.awake = false;

        Body tmpBody;
        synchronized (WorldRender.BOX2D_LOCK) {
            tmpBody = chunk.getWorld().getBox2dWorld().createBody(bodyDef);
        }
        EdgeShape edgeShape = new EdgeShape();
//        synchronized (WorldRender.BOX2D_LOCK) {
        synchronized (this) {
            for (int localX = 0; localX < CHUNK_SIZE; localX++) {
                for (int localY = 0; localY < CHUNK_SIZE; localY++) {
                    @NotNull Block b = chunk.getRawBlock(localX, localY);

                    if (b == null || !b.getMaterial().isSolid()) {
                        continue;
                    }

                    int worldX = CoordUtil.chunkToWorld(chunk.getChunkX(), localX);
                    int worldY = CoordUtil.chunkToWorld(chunk.getChunkY(), localY);

                    for (Tuple<Direction, byte[]> tuple : EDGE_DEF) {
                        Direction dir = tuple.key;

                        //FIXME only check the chunk if the local coordinates are outside this chunk
                        if (!CoordUtil.isInsideChunk(localX + dir.dx, localY + dir.dy) && //
                            !chunk.getWorld().isChunkLoaded(CoordUtil.worldToChunk(worldX + dir.dx),
                                                            CoordUtil.worldToChunk(worldY + dir.dy))) {
                            continue;
                        }

                        Block rel;
                        if (CoordUtil.isInsideChunk(localX + dir.dx, localY + dir.dy)) {
                            rel = chunk.getRawBlock(localX + dir.dx, localY + dir.dy);
                        }
                        else {
                            Chunk relChunk = chunk.getWorld().getChunkFromWorld(worldX + dir.dx, worldY + dir.dy);
                            int relOffsetX = CoordUtil.chunkOffset(worldX + dir.dx);
                            int relOffsetY = CoordUtil.chunkOffset(worldY + dir.dy);
                            rel = relChunk.getBlocks()[relOffsetX][relOffsetY];
                        }
                        if (rel == null || !rel.getMaterial().isSolid() ||
                            (dir == Direction.NORTH && localY == CHUNK_SIZE - 1)) {
                            byte[] ds = tuple.value;
                            edgeShape.set(localX + ds[0], localY + ds[1], localX + ds[2], localY + ds[3]);

                            synchronized (WorldRender.BOX2D_LOCK) {
                                Fixture fix = tmpBody.createFixture(edgeShape, 0);
                                if (!b.getMaterial().blocksLight()) {
                                    fix.setFilterData(World.SOLID_TRANSPARENT_FILTER);
                                }
                            }
                        }
                    }
                }
            }
        }
        edgeShape.dispose();

        if (box2dBody != null) {
            synchronized (WorldRender.BOX2D_LOCK) {
                chunk.getWorld().getBox2dWorld().destroyBody(box2dBody);
            }
        }

        box2dBody = tmpBody;

        Gdx.app.postRunnable(() -> chunk.getWorld().getRender().update());


        boolean potentiallyDirty = false;

        //TODO Try to optimize this (ie select what directions to recalculate)
        for (Direction direction : Direction.CARDINAL) {
            Location relChunk = Location.relative(chunk.getChunkX(), chunk.getChunkY(), direction);
            if (chunk.getWorld().isChunkLoaded(relChunk)) {
                if (recalculateNeighbors) {
                    //noinspection LibGDXFlushInsideLoop
                    chunk.getWorld().getChunk(relChunk).getChunkBody().updateFixture(false);
                }
            }
            else {
                potentiallyDirty = true;
            }
        }

        if (potentiallyDirty) {
            Main.inst().getScheduler().scheduleSync(() -> {
                updateFixture(false);
            }, 10);
        }
    }

    @Override
    public void dispose() {
        if (box2dBody != null) {
            synchronized (WorldRender.BOX2D_LOCK) {
                chunk.getWorld().getBox2dWorld().destroyBody(box2dBody);
            }
            box2dBody = null;
        }
    }
}

File: core/src/main/kotlin/no/elg/infiniteBootleg/world/box2d/ChunkBody.kt
package no.elg.infiniteBootleg.world.box2d

import com.badlogic.gdx.physics.box2d.Body
import com.badlogic.gdx.physics.box2d.BodyDef
import com.badlogic.gdx.physics.box2d.BodyDef.BodyType.StaticBody
import com.badlogic.gdx.physics.box2d.ChainShape
import com.badlogic.gdx.physics.box2d.Fixture
import com.badlogic.gdx.utils.LongMap
import no.elg.infiniteBootleg.api.Updatable
import no.elg.infiniteBootleg.util.CheckableDisposable
import no.elg.infiniteBootleg.util.compactLoc
import no.elg.infiniteBootleg.util.isAir
import no.elg.infiniteBootleg.world.BOX2D_LOCK
import no.elg.infiniteBootleg.world.blocks.Block
import no.elg.infiniteBootleg.world.blocks.Block.Companion.compactWorldLoc
import no.elg.infiniteBootleg.world.blocks.EntityMarkerBlock
import no.elg.infiniteBootleg.world.chunks.Chunk
import no.elg.infiniteBootleg.world.chunks.Chunk.Companion.CHUNK_SIZE
import no.elg.infiniteBootleg.world.ecs.components.events.PhysicsEvent
import no.elg.infiniteBootleg.world.ecs.components.events.PhysicsEventQueue.Companion.queuePhysicsEvent
import no.elg.infiniteBootleg.world.world.World
import javax.annotation.concurrent.GuardedBy

/**
 * @author Elg
 */
class ChunkBody(private val chunk: Chunk) : Updatable, CheckableDisposable {

  private val fixtureMap = LongMap<Fixture>()

  private val chunkBodyLock = Any()

  /**
   * The actual box2d body of the chunk.
   *
   * The setter is locked under [bodyLock], the old body will automatically be destroyed.
   *
   * The getter is **not** locked under [bodyLock]
   */
  @field:Volatile
  private var box2dBody: Body? = null
    set(value) {
      val oldBody: Body?
      synchronized(chunkBodyLock) {
        oldBody = field
        field = value
        fixtureMap.clear()
      }
      if (oldBody != null) {
        // We should now be fine to destroy the old body
        chunk.world.worldBody.destroyBody(oldBody)
      }
    }

  @field:Volatile
  private var disposed = false

  override val isDisposed: Boolean
    get() = disposed

  /**calculate the shape of the chunk (box2d)*/
  val bodyDef = BodyDef().also {
    it.position.set(chunk.chunkX * CHUNK_SIZE.toFloat(), chunk.chunkY * CHUNK_SIZE.toFloat())
    it.fixedRotation = true
    it.type = StaticBody
  }

  /**
   * Update the box2d fixture of this chunk
   *
   * @param recalculateNeighbors
   * If the neighbors also should be updated
   */
  override fun update() {
    chunk.world.worldBody.updateChunk(this)
  }

  @GuardedBy("BOX2D_LOCK")
  fun shouldCreateBody(): Boolean {
    if (isDisposed) {
      return false
    }
    if (chunk.isAllAir) {
      box2dBody = null
      return false
    }
    return true
  }

  /**
   * Must be called under [BOX2D_LOCK]
   *
   * @see World.postBox2dRunnable
   * @see WorldBody.postBox2dRunnable
   */
  @GuardedBy("BOX2D_LOCK")
  fun onBodyCreated(tmpBody: Body) {
    for (localX in 0 until CHUNK_SIZE) {
      for (localY in 0 until CHUNK_SIZE) {
        val block = chunk.getRawBlock(localX, localY)
        if (block.isAir()) {
          continue
        }
        addBlock(block)
      }
    }

    // if this got disposed while creating the new chunk fixture, this is the easiest cleanup solution
    if (isDisposed) {
      box2dBody = null
      chunk.world.worldBody.destroyBody(tmpBody)
    } else {
      box2dBody = tmpBody
    }
  }

  fun removeBlock(block: Block) {
    chunk.world.postBox2dRunnable {
      fixtureMap.get(compactLoc(block.localX, block.localY))?.also { fixture ->
        fixture.filterData = Filters.NON_INTERACTIVE__GROUND_FILTER
        fixture.userData = null
        chunk.world.engine.queuePhysicsEvent(PhysicsEvent.BlockRemovedEvent(fixture, block.compactWorldLoc))
      }
    }
  }

  fun addBlock(block: Block, box2dBody: Body? = null) {
    val worldBody = chunk.world.worldBody
    chunk.world.postBox2dRunnable {
      val body = box2dBody ?: this.box2dBody
      if (body == null) {
        worldBody.updateChunk(this@ChunkBody)
        return@postBox2dRunnable
      }
      val localX = block.localX
      val localY = block.localY

      val compactLoc = compactLoc(localX, localY)
      val cacheFix: Fixture? = fixtureMap.get(compactLoc)
      val fixture: Fixture = if (cacheFix != null) {
        cacheFix
      } else {
        val chainShape = ChainShape()
        chainShape.createLoop(
          floatArrayOf(
            localX + 0f,
            localY + 0f,

            localX + 0f,
            localY + 1f,

            localX + 1f,
            localY + 1f,

            localX + 1f,
            localY + 0f
          )
        )
        val newFix = body.createFixture(chainShape, 0f)
        fixtureMap.put(compactLoc, newFix)

        chainShape.dispose()
        newFix
      }

      val material = block.material
      fixture.userData = block
//      chunk.world.engine.queuePhysicsEvent(PhysicsEvent.BlockChangedEvent(fixture, material))
      fixture.filterData = when {
        block is EntityMarkerBlock -> Filters.NON_INTERACTIVE__GROUND_FILTER
        material.isCollidable -> Filters.GR_FB_EN__GROUND_FILTER
        else -> Filters.GR_FB__GROUND_FILTER
      }
    }
  }

  override fun dispose() {
    if (isDisposed) return
    disposed = true
    box2dBody = null
  }

  override fun equals(other: Any?): Boolean {
    if (this === other) return true
    if (other !is ChunkBody) return false
    return chunk == other.chunk
  }

  override fun hashCode(): Int {
    return chunk.hashCode()
  }
}
